src/ast/passes/semantic_analyser.cpp

```diff
@@ -2751,81 +2745,68 @@ void SemanticAnalyser::visit(AssignVarStatement &assignment)
                 << (integer->is_negative ? integer->n
                                          : static_cast<uint64_t>(integer->n))
                 << "' which does not fit into the variable of type '"
-                << search->second << "'";
+                << storedTy << "'";
           }
         }
-      } else if (search->second.IsSigned() != assignTy.IsSigned()) {
+      } else if (storedTy.IsSigned() != assignTy.IsSigned()) {
         LOG(ERROR, assignment.loc, err_)
             << "Type mismatch for " << var_ident << ": "
             << "trying to assign value of type '" << assignTy
-            << "' when variable already contains a value of type '"
-            << search->second << "'";
+            << "' when variable already contains a value of type '" << storedTy
+            << "'";
       } else {
-        if (!assignTy.FitsInto(search->second)) {
+        if (!assignTy.FitsInto(storedTy)) {
           LOG(ERROR, assignment.loc, err_)
               << "Integer size mismatch. Assignment type '" << assignTy
-              << "' is larger than the variable type '" << search->second
-              << "'.";
+              << "' is larger than the variable type '" << storedTy << "'.";
+        }
+      }
+    } else if (assignTy.IsRecordTy()) {
+      if (assignTy.GetName() != storedTy.GetName()) {
+        LOG(ERROR, assignment.loc, err_)
+            << "Type mismatch for " << var_ident << ": "
+            << "trying to assign value of type '" << assignTy.GetName()
+            << "' when variable already contains a value of type '" << storedTy
+            << "'";
+      }
+    } else if (assignTy.IsBufferTy()) {
+      auto var_size = storedTy.GetSize();
+      auto expr_size = assignTy.GetSize();
+      if (var_size != expr_size) {
+        LOG(WARNING, assignment.loc, out_)
+            << "Buffer size mismatch: " << var_size << " != " << expr_size
+            << (var_size < expr_size ? ". The value may be truncated."
+                                     : ". The value may contain garbage.");
+      }
+    } else if (assignTy.IsTupleTy()) {
+      update_string_size(storedTy, assignTy);
+      // Early passes may not have been able to deduce the full types of tuple
+      // elements yet. So wait until final pass.
+      if (is_final_pass()) {
+        if (!assignTy.FitsInto(storedTy)) {
+          LOG(ERROR, assignment.loc, err_)
+              << "Tuple type mismatch: " << storedTy << " != " << assignTy
+              << ".";
         }
       }
     }
   } else {
     // This variable hasn't been seen before
-    variable_val_[scope_].insert({ var_ident, assignment.expr->type });
+    variable_val_[scope_].insert({ var_ident, assignTy });
   }
 
-  auto &storedTy = variable_val_[scope_][var_ident];
-
+  const auto &storedTy = variable_val_[scope_][var_ident];
   assignment.var->type = storedTy;
 
-  if (assignTy.IsRecordTy()) {
```

Comment:
mostly just moved this into the block above when we already have a previous type for the var